test/e2e/memtierd.test-suite/pidwatcher/n4c16/test02-proc/code.var.sh

```diff
@@ -0,0 +1,206 @@
+vm-command "[ -f /sys/kernel/mm/page_idle/bitmap ]" || {
+    if [[ "$distro" != "debian-sid" && "$distro" != "ubuntu-22.04" ]]; then
+        error "idlepage e2e test is implemented only for distro=debian-sid or distro=ubuntu-22.04"
+    fi
+
+    if [ "$distro" == "debian-sid" ]; then
+        damon-idlepage-setup
+    fi
+
+    vm-command "[ -f /sys/kernel/mm/page_idle/bitmap ]" || error "failed to setup idlepage"
+}
+
+memtierd-setup
+
+# Create swap
+zram-install
+zram-swap off
+zram-swap 2G
+
+# start meme
+for i in {0..2}; do
+    MEME_BS=1G MEME_BWC=1 MEME_BWS=300M MEME_MEMS=0 memtierd-meme-start
+    PID_VAR="MEME${i}_PID"
+    eval "${PID_VAR}=\$MEME_PID"
+done
+
+next-round() {
+    local round_counter_var=$1
+    local round_counter_val=${!1}
+    local round_counter_max=$2
+    local round_delay=$3
+    if [[ "$round_counter_val" -ge "$round_counter_max" ]]; then
+        return 1
+    fi
+    eval "$round_counter_var=$(($round_counter_val + 1))"
+    sleep $round_delay
+    return 0
+}
+
+match-pids() {
+    for pid_regexp in "$@"; do
+      round_number=0
+      while ! ( memtierd-command "stats -t memory_scans -f csv | awk -F, \"NR>3 {print \\\$1}\""; grep ${pid_regexp} <<< $COMMAND_OUTPUT); do
+          echo "grep pid matching ${pid_regexp} not found"
+          next-round round_number 5 1 || {
+              error "timeout: memtierd did not watch ${pid_regexp}"
+          }
+      done
+    done
+}
+
+echo -e "\n=== scenario 1: test policy-age and tracker-softdirty ===\n"
+MEMTIERD_YAML="
+policy:
+  name: age
+  config: |
+    intervalms: 4000
+    pidwatcher:
+      name: filter
+      config: |
+        source:
+          name: proc
+          config: |
+            intervalms: 4000
+        filters:
+          - procexeregexp: ".*/meme"
+    idledurationms: 8000
+    idlenumas: [3]
+    activedurationms: 6000
+    activenumas: [1]
+    tracker:
+      name: softdirty
+      config: |
+        pagesinregion: 256
+        maxcountperregion: 0
+        scanintervalms: 4000
+        regionsupdatems: 0
+        skippageprob: 0
+        pagemapreadahead: 0
+    mover:
+      intervalms: 20
+      bandwidth: 200
+"
+memtierd-start
+
+sleep 5
+match-pids $MEME0_PID $MEME1_PID $MEME2_PID
+
```

Comment:
In case of proc + filter pidwatcher combination it would be good to check that these are exactly the pids that are tracked. That is, the filter should be fitering out pids of all other processes.

Reply:
zram/swap is not really needed.
==> Done and deleted.

It suffices to test one tracker per policy, there is no need to test all combinations of pidwatchers, trackers and policies.
It would be nice to test that the multitracker properly passes found pids to all trackers.
==> So now the following scenarios will be covered as we have discussed: 
- Age policy with multiple-tracker (softdirty and idlepage): test pidlist, cgroups, filter with proc, filter with cgroups
- Age policy with multiple-tracker (softdirty and idlepage): test pidlist, cgroups, filter with proc, filter with cgroups

Would it be possible include the damon tracker, too?
==> After checking with you, daemon tracker might be problematic, so maybe wont add it in this pr.

In case of proc + filter pidwatcher combination it would be good to check that these are exactly the pids that are tracked. That is, the filter should be fitering out pids of all other processes.
==> yeah, I will improve match-pids function.

Reply:
Done